components/org.wso2.carbon.identity.application.authenticator.google/src/main/java/org/wso2/carbon/identity/application/authenticator/google/GoogleOAuth2Authenticator.java

```diff
@@ -58,6 +70,158 @@ private void initTokenEndpoint() {
+        validCredential =  idToken != null ? true:false;
+        return validCredential;
+    }
+
+    /**
+     * This function validates the CSRF double-sided cookie returned from Google One Tap respond
+     * The request is considered as non-attacked request if the CSRF cookie and the parameter is equal
+     * @param request
+     * @return
+     */
+    private boolean validateCSRFCookies(HttpServletRequest request) {
+
+        boolean validCookies =  false;
+
+        if (request.getCookies() != null) {
+
+            String crossRefCookieHalf;
+            String crossRefParamHalf = request.getParameter("g_csrf_token");
+
+            List<Cookie> crossRefCookies = Arrays.stream(request.getCookies())
+                    .filter(cookie -> cookie.getName().equalsIgnoreCase("g_csrf_token"))
+                    .collect(Collectors.toList());
+
+            if (crossRefCookies == null || crossRefCookies.isEmpty() || crossRefCookies.get(0) == null) {
+                if (log.isDebugEnabled()) {
+                    log.debug("No CSRF cookie found. Invalid request");
+                }
+            } else {
+                crossRefCookieHalf = crossRefCookies.get(0).getValue();
+
+                if (crossRefParamHalf == null || crossRefParamHalf.isEmpty() || crossRefCookieHalf.isEmpty()) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("No CSRF parameter found. Invalid request");
+                    }
+                } else if (!crossRefParamHalf.equals(crossRefCookieHalf)) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("CSRF validation failed for Google One Tap");
+                    }
+                } else {
+                    validCookies = true;
+                }
+            }
+        }
+        return validCookies;
+    }
+
+    @Override
+    protected String mapIdToken(AuthenticationContext context, HttpServletRequest request, OAuthClientResponse oAuthResponse) {
+        String idToken;
+        if (StringUtils.isNotBlank(request.getParameter("oneTapEnabled")) &&
+                request.getParameter("oneTapEnabled").equals("YES")){
+             idToken = request.getParameter("credential");
+            context.setProperty(OIDCAuthenticatorConstants.ID_TOKEN, idToken);
+        } else {
+            idToken = super.mapIdToken(context, request, oAuthResponse);
+        }
+        return idToken;
+    }
+
+    @Override
+    protected boolean isInitialRequest(AuthenticationContext context, HttpServletRequest request) {
+        boolean isInitialRequest;
+        if (StringUtils.isNotBlank(request.getParameter("oneTapEnabled")) &&
+                request.getParameter("oneTapEnabled").equals("YES")){
+            isInitialRequest = false;
+            context.setCurrentAuthenticator(getName());
+            context.setProperty("oneTapEnabled", request.getParameter("oneTapEnabled").equalsIgnoreCase("YES"));//Setting onetap config at context level
```

Comment:
Shall we move this comment to a newline? In that case, we can easily view the comment in the editor without scrolling.

Reply:
Fixed 44ae7456fe0af189a80e5ee90a90781fec1d2386